### PR TITLE
Use Cargo's target directory sharing.

### PR DESCRIPTION
**Pull request by metajack**
This speeds up `./mach build --dev` followed by `./mach build-cef` by
25%. When rust-lang/cargo#497 is fixed, this speedup will increase
dramatically.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6306)

### CODE REVIEW

File: python/servo/command_base.py [hunk 1]
```diff
@@ -130,8 +130,14 @@ def cargo_build_id(self):
...
 
+    def get_target_dir(self):
+        if "CARGO_TARGET_DIR" in os.environ:
+            return os.environ["CARGO_TARGET_DIR"]
+        else:
+            return path.join(self.context.topdir, "target")
+
     def get_binary_path(self, release, dev):
-        base_path = path.join("components", "servo", "target")
+        base_path = self.get_target_dir()
```
Comment (mbrubeck) [inline python/servo/command_base.py]: `build_commands.py`, `post_build_commands.py`, and `testing_commands.py` still contain references to the old target directory.

**Comment by SimonSapin** [comment]
Review status: 0 of 4 files reviewed, 1 unresolved discussion, all commit checks successful.

---

<sup>**[python/servo/command_base.py, line 140 \[r1\]](https://reviewable.io:443/reviews/servo/servo/6306#-JrKDTWudWDWASxKdC9M-r1-140)** ([raw file](https://github.com/servo/servo/blob/87df1886af7a8de14e525d1a54c04591019ec599/python/servo/command_base.py#L140)):</sup>
Also:

```
components/util/resource_files.rs
etc/ci/upload_docs.sh
ports/gonk/README.md
python/servo/command_base.py
python/tidy.py
tests/power/PowerMeasure.py
tests/wpt/harness/test/test.cfg.example

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)
<!-- Sent from Reviewable.io -->

**Comment by mbrubeck** [comment]
Review status: all files reviewed, 1 unresolved discussion, some commit checks pending.
Reviewed files:
- cargo-nightly-build @ r1
- components/util/resource_files.rs @ r2
- etc/ci/upload_docs.sh @ r2
- ports/cef/Cargo.toml @ r1
- ports/gonk/README.md @ r2
- python/servo/build_commands.py @ r2
- python/servo/command_base.py @ r1
- python/servo/post_build_commands.py @ r2
- python/servo/testing_commands.py @ r2
- python/tidy.py @ r1
- tests/power/PowerMeasure.py @ r2
- tests/wpt/harness/test/test.cfg.example @ r2

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)

**Comment by metajack** [comment]
I missed run.py for wpt. Fixed that.

re-r? @SimonSapin or @mbrubeck

**Comment by mbrubeck** [comment]
Review status: all files reviewed, 1 unresolved discussion, all commit checks successful.
Reviewed files:
- tests/wpt/run.py @ r3

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)

**Comment by mbrubeck** [comment]
`./mach test-unit` failed compiling gfx_tests with `clang: error: linker command failed with exit code 1 (use -v to see invocation)`.  I'm not sure why.

**Comment by metajack** [comment]
Blocked on https://github.com/rust-lang/cargo/issues/1695

**Comment by mbrubeck** [comment]
Review status: all files reviewed, 1 unresolved discussion, all commit checks successful.
Reviewed files:
- cargo-nightly-build @ r5

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)

**Comment by metajack** [comment]
Never seen this error before.

**Comment by metajack** [comment]
@acrichto any idea about the above cargo error?

**Comment by metajack** [comment]
@alexcrichton will probably work better :)

**Comment by mbrubeck** [comment]
The error is only happening on our Android cross-compile builds.  These use the same Cargo.toml and Cargo.lock as our regular builds, but maybe something is different in the environment variables or in some platform-specific dependencies...

**Comment by mbrubeck** [comment]
The assertion is failing because `package_ids.len() == 111`, but `ret.len() == 112`.  `ret` includes two entries for `"android_glue"`.

**Comment by mbrubeck** [comment]
The two PackageIds for `android_glue` differ in the `src_path` field:

```
/home/servo/buildbot/slave/android/build/support/android-rs-glue/glue/src/lib.rs
/home/servo/buildbot/slave/android/build/components/servo/../../support/android-rs-glue/glue/src/lib.rs
```

and in the same way in the `manifest_path` field.

**Comment by alexcrichton** [comment]
Gah sorry this cargo upgrade has been so much of a pain :(. Investigating to see what's going on, but thanks for the quick diagnosis @mbrubeck!

**Comment by alexcrichton** [comment]
Hm, so [removing this line](https://github.com/servo/servo/blob/master/components/servo/.cargo/config#L1) removes that failure, although I'm still investigating why that line causes the bug. How come that exists?

**Comment by mbrubeck** [comment]
> Hm, so [removing this line](https://github.com/servo/servo/blob/master/components/servo/.cargo/config#L1) removes that failure, although I'm still investigating why that line causes the bug. How come that exists?

We have an in-tree copy of android-rs-glue because we need to build its `apk-builder` crate before running `cargo build`.  I think that override exists just to avoid downloading a second copy of the source to build the `android_glue` crate.

**Comment by larsbergstrom** [comment]
@alexcrichton because otherwise we don't know the concrete path to use in: https://github.com/servo/servo/blob/master/components/servo/.cargo/config#L4

**Comment by alexcrichton** [comment]
Cargo bug fixed in https://github.com/rust-lang/cargo/pull/1697, hopefully it's the last one!

**Comment by alexcrichton** [comment]
er sorry, fixed in https://github.com/rust-lang/cargo/pull/1709

**Comment by mbrubeck** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)

**Comment by mbrubeck** [comment]
Compiling servo v0.0.1 (file:///home/servo/buildbot/slave/android/build/components/servo)
    error: could not exec the linker `../../support/android-rs-glue/apk-builder/target/debug/apk-builder`: No such file or directory (os error 2)
    Could not compile `servo`.

**Comment by mbrubeck** [comment]
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6306)

**Comment by alexcrichton** [comment]
:confetti_ball:

**Pull request merged**